File: C++/Lab/DT/Lab1.cpp
#include <iostream>
using namespace std;

void cmpBinIter(int *data);
int iterSearch(int *data, int size, int num);
int recSearch(int *data, int start_idx, int size, int num);
void recAdd(int n, int m);
void recDiff(int n, int m);
void recSum(int k);
int main()
{
    int raw_data[10] = {10, 20, 30, 40, 50, 60, 70, 80, 90, 100};
    cmpBinIter(raw_data);
    cout << "\n";

    int num1, num2;
    cout << "Enter first number : ";
    cin >> num1;
    cout << "Enter second number : ";
    cin >> num2;
    recAdd(num1, num2);
    recDiff(num1, num2);

    int num3;
    cout << "Enter a number : ";
    cin >> num3;
    cout << "Result is : ";
    recSum(num3);
}

void cmpBinIter(int *data)
{
    int target;
    int choice;
    while (true)
    {
        cout << "Enter an integer to search : ";
        cin >> target;
        cout << "Enter an method of search(1. Binary\t2. Recursive Binary) : ";
        cin >> choice;

        int position;
        switch (choice)
        {
            case 1:
                position = iterSearch(data, 10, target);
                break;
            case 2:
                position = recSearch(data, 0, 9, target);
                break;
            default:
                cout << "Wrong command." << endl;
                continue;
        }

        if (position == -1)
            {
                cout << target << " is NOT FOUND.\n";
                break;
            }
        else
            cout << target << " is at position " << position << endl;
    }
}
int iterSearch(int *data, int size, int target)
{
    int low_idx = 0, high_idx = size - 1;
    while (low_idx <= high_idx)
    {
        int mid_idx = (low_idx + high_idx) / 2;

        if (data[mid_idx] == target)
            return mid_idx;
        else if (data[mid_idx] > target)
            high_idx = mid_idx - 1;
        else
            low_idx = mid_idx + 1;
    }
    return -1;
}
int recSearch(int *data, int low_idx, int high_idx, int target)
{
    int mid_idx = (low_idx + high_idx) / 2;
    if (low_idx <= high_idx)
    {
        if (data[mid_idx] == target)
            return mid_idx;
        else if (data[mid_idx] > target)
            return recSearch(data, low_idx, mid_idx - 1, target);
        else
            return recSearch(data, mid_idx + 1, high_idx, target);
    }
    else
        return -1;
}

void recAdd(int n, int m)
{
    if (n == 0)
        cout << "Addition Result is : " << m << endl;
    else
        recAdd(n-1, m+1);
}
void recDiff(int n, int m)
{
    if (n == 0)
        cout << "Difference Result is : " << m << endl;
    else
        recDiff(n-1, m-1);
}
void recSum(int k)
{
    static int sum = 0;
    sum += k;

    if (k == 1)
        cout << k << " = " << sum << endl;
    else
    {
        cout << k << " + "; 
        recSum(k-1);
    }
}

File: Python3/Lab/Lab1.py
# 20181670, 이재욱, Lab2

def binarySearchIteration(data, target):
    if len(data) == 0:
        return -1

    left_idx = 0
    right_idx = len(data) - 1
    status = False

    while left_idx <= right_idx:
        mid_idx = (left_idx + right_idx) // 2

        if data[mid_idx] == target:
            status = True
            return mid_idx

        elif data[mid_idx] > target:
            right_idx = mid_idx - 1
        else:
            left_idx = mid_idx + 1

    if not status:
        return -1


def binarySearchRecursive(data, target, left, right):
    if len(data) == 0:
        return -1

    if left <= right:
        mid_idx = (left + right) // 2

        if data[mid_idx] == target:
            status = True
            return mid_idx
        elif data[mid_idx] > target:
            return binarySearchRecursive(data, target, left, mid_idx - 1)
        else:
            return binarySearchRecursive(data, target, mid_idx + 1, right)
    else:
        return -1


def recursiveSum(n, m):
    if n == 0:
        return n + m
    else:
        n -= 1
        m += 1
        return recursiveSum(n, m)


def recursiveDiff(n, m):
    if n == 0:
        return n - m
    else:
        n -= 1
        m -= 1
        return recursiveDiff(n, m)


def recursiveRangeSum(n):
    if n == 1:
        return 1
    else:
        return n + recursiveRangeSum(n-1)


if __name__ == '__main__':
    # Lab2 - 1
    list = [10, 20, 30, 40, 50, 60, 70, 80, 90, 100]
    while True:
        find_number = int(input("Enter an integer to search : "))
        method_choose = int(input("Enter method of search : (1. Binary Search   2. Recursive Search) : "))
        position = -1 # Initiate position '-1' to break the while statement

        if method_choose == 1:
            position = binarySearchIteration(list, find_number)

            if position == -1:
                print("%d is NOT FOUND" %find_number)
            else:
                print("%d is at position %d" %(find_number, position))

        elif method_choose == 2:
            position = binarySearchRecursive(list, find_number, 0, len(list)-1)

            if position == -1:
                print("%d is NOT FOUND" % find_number)
            else:
                print("%d is at position %d" % (find_number, position))
        else:
            print("Wrong order!")

        if position == -1:
            break

    # Lab2 - 2
    a = int(input("Enter number1 : "))
    b = int(input("Enter number2 : "))
    print("Addition Result is : %d" %recursiveSum(a, b))
    print("Difference Result : %d" %recursiveDiff(a, b))

    c = int(input("Enter a number : "))
    equation = ""
    for i in range(c, 0, -1):
        if i == 1:
            equation += "%d = " %i
        else:
            equation += "%d + " %i
    print("Result is : %s%d" %(equation, recursiveRangeSum(c)))